x-pack/metricbeat/module/azure/monitor/azure_monitor_service.go

```diff
@@ -0,0 +1,91 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package monitor
+
+import (
+	"context"
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/services/preview/monitor/mgmt/2019-06-01/insights"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-03-01/resources"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+type AzureMonitorService struct {
+	metricsClient          *insights.MetricsClient
+	metricDefinitionClient *insights.MetricDefinitionsClient
+	resourceClient         *resources.Client
+	log                    *logp.Logger
+}
+
+// Init instantiates the an Azure monitoring service
+func Init(clientID string, clientSecret string, tenantID string, subscriptionID string) (*AzureMonitorService, error) {
+	clientConfig := auth.NewClientCredentialsConfig(clientID, clientSecret, tenantID)
+	authorizer, err := clientConfig.Authorizer()
+	if err != nil {
+		return nil, err
+	}
+	metricsClient := insights.NewMetricsClient(subscriptionID)
+	metricsDefinitionClient := insights.NewMetricDefinitionsClient(subscriptionID)
+	resourceClient := resources.NewClient(subscriptionID)
+	metricsClient.Authorizer = authorizer
+	metricsDefinitionClient.Authorizer = authorizer
+	resourceClient.Authorizer = authorizer
+	client := &AzureMonitorService{
+		metricDefinitionClient: &metricsDefinitionClient,
+		metricsClient:          &metricsClient,
+		resourceClient:         &resourceClient,
+		log:                    logp.NewLogger("azure monitor"),
+	}
+	return client, nil
+}
+
+// GetResourceById will retrieve the azure resource details based on the id entered
+func (client AzureMonitorService) GetResourceById(resourceID string) (resources.GenericResource, error) {
+	resource, err := client.resourceClient.GetByID(context.Background(), resourceID)
+	if err != nil {
+		client.log.Errorf(" error while retrieving resource by id  %s : %v", resourceID, err)
+	}
+	return resource, err
+}
+
+// GetResourcesByResourceGroup will retrieve the list of resources that match the resource group and type
+func (client *AzureMonitorService) GetResourcesByResourceGroup(resourceGroup string, resourceType string) ([]resources.GenericResource, error) {
+	var top int32 = 500
+	result, err := client.resourceClient.ListByResourceGroup(context.Background(), resourceGroup, fmt.Sprintf("resourceType eq '%s'", resourceType), "true", &top)
+	if err != nil {
+		client.log.Errorf("error while listing resources by resource group %s  and filter %s : %s", resourceGroup, resourceType, err)
+	}
+	return result.Values(), err
```

Comment:
in case of error, isn't result nil?

Reply:
I implemented the error handling, this is old code